dpcpp/matrix/csr_kernels.dp.cpp

```diff
@@ -71,11 +78,1812 @@ namespace dpcpp {
+void abstract_merge_path_spmv(
+    const IndexType num_rows, const ValueType *__restrict__ alpha,
+    const ValueType *__restrict__ val, const IndexType *__restrict__ col_idxs,
+    const IndexType *__restrict__ row_ptrs, const IndexType *__restrict__ srow,
+    const ValueType *__restrict__ b, const size_type b_stride,
+    const ValueType *__restrict__ beta, ValueType *__restrict__ c,
+    const size_type c_stride, IndexType *__restrict__ row_out,
+    ValueType *__restrict__ val_out, sycl::nd_item<3> item_ct1,
+    IndexType *shared_row_ptrs)
+{
+    const auto alpha_val = alpha[0];
+    const auto beta_val = beta[0];
+    merge_path_spmv<items_per_thread>(
+        num_rows, val, col_idxs, row_ptrs, srow, b, b_stride, c, c_stride,
+        row_out, val_out, [&alpha_val](ValueType &x) { return alpha_val * x; },
+        [&beta_val](ValueType &x) { return beta_val * x; }, item_ct1,
+        shared_row_ptrs);
+}
+
+template <int items_per_thread, typename ValueType, typename IndexType>
+void abstract_merge_path_spmv(
+    dim3 grid, dim3 block, size_type dynamic_shared_memory, sycl::queue *queue,
+    const IndexType num_rows, const ValueType *alpha, const ValueType *val,
+    const IndexType *col_idxs, const IndexType *row_ptrs, const IndexType *srow,
+    const ValueType *b, const size_type b_stride, const ValueType *beta,
+    ValueType *c, const size_type c_stride, IndexType *row_out,
+    ValueType *val_out)
+{
+    queue->submit([&](sycl::handler &cgh) {
+        sycl::accessor<IndexType, 1, sycl::access_mode::read_write,
+                       sycl::access::target::local>
+            shared_row_ptrs_acc_ct1(
+                sycl::range<1>(spmv_block_size * items_per_thread), cgh);
+
+        cgh.parallel_for(
+            sycl_nd_range(grid, block), [=](sycl::nd_item<3> item_ct1) {
+                abstract_merge_path_spmv<items_per_thread>(
+                    num_rows, alpha, val, col_idxs, row_ptrs, srow, b, b_stride,
+                    beta, c, c_stride, row_out, val_out, item_ct1,
+                    static_cast<IndexType *>(
+                        shared_row_ptrs_acc_ct1.get_pointer()));
+            });
+    });
+}
+
+
+template <typename ValueType, typename IndexType>
+void abstract_reduce(const IndexType nwarps,
+                     const ValueType *__restrict__ last_val,
+                     const IndexType *__restrict__ last_row,
+                     ValueType *__restrict__ c, const size_type c_stride,
+                     sycl::nd_item<3> item_ct1,
+                     UninitializedArray<IndexType, spmv_block_size> *tmp_ind,
+                     UninitializedArray<ValueType, spmv_block_size> *tmp_val)
+{
+    merge_path_reduce(
+        nwarps, last_val, last_row, c, c_stride, [](ValueType &x) { return x; },
+        item_ct1, tmp_ind, tmp_val);
+}
+
+template <typename ValueType, typename IndexType>
+void abstract_reduce(dim3 grid, dim3 block, size_type dynamic_shared_memory,
```

Comment:
Do you actually use this `dynamic_shared_memory` parameter anywhere? You could consider removing it wherever it's not used, which seems to be everywhere.

Reply:
I can only remove the variable name to avoid the warning.
I can not remove the variable because it needs to provide a consistent kernel call.
`kernel(grid, block, dynamic_shared_memory, queue, ...)`

Reply:
Is that because your script produces these functions, and it needs a common function signature?

Reply:
function signature is for calling.
when we have the kernel, user will not know the function contains grid, block, dynamic, queue
like hiplaunchkernel always has these four args

Reply:
So you are planning to have something like a common `syclLaunchKernel`? Because if not, each of these functions can then have its own signature, right?

Reply:
Yes, each of these functions can have its own signature.
the signature what I mentioned is mainly for kernel syntax.
in cuda, the syntax is in `<<<` `>>>`
in hip, the first five args are syntax in `hipLaunchKernelGGL(kernel_name, grid, block, dynamic_shared_memory, stream, others args)`
in dpcpp, I would like have something similar like this. the first four arguments are kernel syntax